applications/matter_bridge/src/bridged_devices_creator.cpp

```diff
@@ -0,0 +1,384 @@
+	chip::DeviceLayer::PlatformMgr().ScheduleWork(
+		[](intptr_t context) {
+			BluetoothConnectionContext *ctx = reinterpret_cast<BluetoothConnectionContext *>(context);
+			if (CHIP_NO_ERROR != AddMatterDevice(ctx->deviceType, ctx->nodeLabel, ctx->provider, ctx->index,
+							     ctx->endpointId)) {
+				BLEConnectivityManager::Instance().RemoveBLEProvider(ctx->address);
+			}
+		},
+		reinterpret_cast<intptr_t>(ctx));
+	return;
+
+exit:
+	BLEConnectivityManager::Instance().RemoveBLEProvider(ctx->address);
+}
+
+BridgedDeviceDataProvider *CreateBleProvider(int deviceType)
+{
+	return BridgeFactory::GetBleDataProviderFactory().Create(
+		static_cast<MatterBridgedDevice::DeviceType>(deviceType), BridgeManager::HandleUpdate);
+}
+#endif /* CONFIG_BRIDGED_DEVICE_BT */
+} /* namespace */
+
+#ifdef CONFIG_BRIDGED_DEVICE_BT
+CHIP_ERROR BridgedDeviceCreator::CreateDevice(int deviceType, const char *nodeLabel, bt_addr_le_t btAddress,
+					      chip::Optional<uint8_t> index, chip::Optional<uint16_t> endpointId)
+{
+	CHIP_ERROR err;
+
+	/* Check if there is already existing provider for given address.
+	 * In case it is, the provider was either connected or recovered before. All what needs to be done is creating
+	 * new Matter endpoint. In case there isn't, the provider has to be created.
+	 */
+	BLEBridgedDeviceProvider *provider = BLEConnectivityManager::Instance().FindBLEProvider(btAddress);
+	if (provider) {
+		return AddMatterDevice(deviceType, nodeLabel, provider, index, endpointId);
+	}
+
+	/* The EnvironmentalSensor is a dummy device type, so it is not saved in the persistent storage. For now the
+	 * only BLE service handling temperature and humidity device types is environmental sensor, so it's necessary to
+	 * perform mapping. */
+	if (deviceType == MatterBridgedDevice::DeviceType::TemperatureSensor ||
+	    deviceType == MatterBridgedDevice::DeviceType::HumiditySensor) {
+		provider = static_cast<BLEBridgedDeviceProvider *>(
+			CreateBleProvider(MatterBridgedDevice::DeviceType::EnvironmentalSensor));
+	} else {
+		provider = static_cast<BLEBridgedDeviceProvider *>(CreateBleProvider(deviceType));
+	}
+
+	if (!provider) {
+		return CHIP_ERROR_NO_MEMORY;
+	}
+
+	err = BLEConnectivityManager::Instance().AddBLEProvider(provider);
+	VerifyOrExit(err == CHIP_NO_ERROR, );
+
+	/* There are two cases for creating a new device:
+	 * - connecting new device for the first time - the index and endpoint id are not relevant, the new Bluetooth LE
+	 * connection has to be created and the operation confirmed by connected callback.
+	 * - recovering a device that was connected in the past after reboot - the index and endpoint id are read from
+	 * storage, the device has to be re-connected without confirming its existence by callback.
+	 */
+	if (index.HasValue() && endpointId.HasValue()) {
+		provider->InitializeBridgedDevice(btAddress, nullptr, nullptr);
+		BLEConnectivityManager::Instance().Recover(provider);
```

Comment:
It's intentional. I think that creating a device for the first time and re-creating it should be handled in a different way. It is very similar to e.g. what we have in Matter regarding connection establishment:

- Once the connection to the device is requested for the first time we basically do not know multiple things. Whether the device with such address even exists, does it support required profiles, in the future will it accept authentication material we have and etc. From this perspective it is not reasonable to create a Matter device, before we know there will be something valid on the other side to connect. In Matter it works in analogous way, we create Matter device instance on the controller only if all stages of commissioning succeeded, otherwise the device is not added.
- Once the re-connection to the device is requested after reboot, we know the device used to exist and we were able to reach it. Moreover we managed to add it and it was e.g. visible in the customer's application. Even if we will not be able to re-connect, we should create a Matter device and set it's status to offline. This way the customer will know he had a device before, but now for some reason it's offline. Otherwise the devices would disappear from the customers UI once it would be not reachable after reboot, what would be super confusing. Again, in Matter we handle re-connects in the same way. If after controllers reboot the device is not reachable, we present it's status as offline.

Reply:
Ok, totally agree (assuming we can indicate that the device is offline).